Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,262 @@
+    public var parent: WorkflowSession? {
+        switch _indirectParent {
+        case .some(let parent):
+            return parent
+        case .none:
+            return nil
+        }
+    }
+
+    init<WorkflowType: Workflow>(
+        workflow: WorkflowType,
+        renderKey: String,
+        parent: WorkflowSession?
+    ) {
+        self.workflowType = WorkflowType.self
+        self.renderKey = renderKey
+        self._indirectParent = IndirectParent(parent)
+    }
+}
+
+// MARK: - No-op Defaults
+
+public extension WorkflowObserver {
+    func sessionDidBegin(
+        _ session: WorkflowSession
+    ) {}
+
+    func sessionDidEnd(
+        _ session: WorkflowSession
+    ) {}
+
+    func workflowDidMakeInitialState<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        initialState: WorkflowType.State,
+        session: WorkflowSession
+    ) {}
+
+    func workflowWillRender<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)? { nil }
+
+    func workflowDidChange<WorkflowType: Workflow>(
+        from oldWorkflow: WorkflowType,
+        to newWorkflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) {}
+
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
+        session: WorkflowSession
+    ) {}
+}
+
+// MARK: Chained Observer
+
+/// 'Chained' observer implementation, which multiplexes a list of `WorkflowObservers` into a
+/// single observer interface.
+final class ChainedWorkflowObserver: WorkflowObserver {
+    private let observers: [WorkflowObserver]
+
+    init(observers: [WorkflowObserver]) {
+        self.observers = observers
+    }
+
+    func sessionDidBegin(_ session: WorkflowSession) {
+        for observer in observers {
+            observer.sessionDidBegin(session)
+        }
+    }
+
+    func sessionDidEnd(_ session: WorkflowSession) {
+        for observer in observers {
+            observer.sessionDidEnd(session)
+        }
+    }
+
+    func workflowDidMakeInitialState<WorkflowType>(_ workflow: WorkflowType, initialState: WorkflowType.State, session: WorkflowSession) where WorkflowType: Workflow {
+        for observer in observers {
+            observer.workflowDidMakeInitialState(workflow, initialState: initialState, session: session)
+        }
+    }
+
+    func workflowWillRender<WorkflowType>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)? where WorkflowType: Workflow {
+        let callbacks = observers.compactMap {
+            $0.workflowWillRender(workflow, state: state, session: session)
+        }
+
+        guard !callbacks.isEmpty else {
+            return nil
+        }
+
+        return { rendering in
+            for callback in callbacks.reversed() {
```

Comment:
`workflowWillRender` is called in order, but the callbacks are called in reverse order. was that intentional?

Reply:
yeah, maybe that's weird (also the behavior should be documented if we keep it). the motivation was the performance tracing use case – you want the start time to be recorded first, and the end time to happen after everything else is done, including any other observers running.

Reply:
ah yep that makes sense!